File: Codechef/MANGOMKT.cpp
#include <bits/stdc++.h>
#include <cstdio>

using namespace std;

#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define trav(a, x) for(auto& a : x)
#define all(x) x.begin(), x.end()
#define sz(x) (int)(x).size()
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;
#define ll long long
#define mp make_pair
#define f first
#define s second

void setIO(string s) {
	freopen((s + ".txt").c_str(), "r", stdin);
	freopen((s + "_out.txt").c_str(), "w", stdout);
}
class mycomparison
{
  bool reverse;
public:
  mycomparison(const bool& revparam=false)
    {reverse=revparam;}
  bool operator() (const pair<int, ll> & lhs, const pair<int, ll> &rhs) const
  {
    if (reverse) return (lhs.s>rhs.s);
    else return (lhs.s<rhs.s);
  }
};
bool fun(pair<int, ll> a, pair<int, ll> b ){
  return a.s < b.s;
}

ll set_query(vector<ll> & a, map<int, set<int> > m) {
  set<int> visited;
  ll ans = 0;
  while(visited.size() < a.size()) {
    priority_queue<pair<int, ll>, vector <pair<int, ll>  >, mycomparison> pq;
    trav(elem, m) {
      if(visited.count(elem.f) == 0){
        pq.push(mp(elem.f, (long long)(elem.s.size()) * (long long)(a[elem.f - 1])));
      }
    }
    pair<int, ll> temp (pq.top());
    ans += a[temp.f - 1];
    visited.insert(temp.f);
    rep(i, 0 , a.size()) {
      if(i + 1 != temp.f) {
        if(m[temp.f].count(i + 1) != 0) {
          a[i] += 1;
        }
        else {
          a[i] -= 1;
        }
      }
    }

  }
  return ans;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);
  int n, m_;
  cin >> n >> m_;
  vector <ll> a(n);
  rep(i,0,n) { cin >> a[i];}
  map<int, set<int> > m;
  rep(i,0,m_){
    int temp1, temp2;
    cin >> temp1 >> temp2;
    if (m.count(temp1) == 0) {
      m[temp1] = set<int>();
    }
    if (m.count(temp2) == 0) {
      m[temp2] = set<int>();
    }
    m[temp1].insert(temp2);
    m[temp2].insert(temp1);
  }
  int querys;
  cin >> querys;
  rep(i,0,querys) {
    string q_type;
    cin >> q_type;
    if(q_type == "+") {
      int temp1, temp2;
      cin >> temp1 >> temp2;
      if (m.count(temp1) == 0) {
        m[temp1] = set<int>();
      }
      if (m.count(temp2) == 0) {
        m[temp2] = set<int>();
      }
      m[temp1].insert(temp2);
      m[temp2].insert(temp1);
    }
    else if (q_type == "-"){
      int temp1, temp2;
      cin >> temp1 >> temp2;
      if (m.count(temp1) == 0 ){
        m[temp1] = set<int>();
      }
      if (m.count(temp2) == 0 ){
        m[temp2] = set<int>();
      }
      m[temp1].erase(temp2);
      m[temp2].erase(temp1);
    }
    else {
      cout << set_query(a, m) << endl;
    }
  }
}

File: Codechef/MANGOMKT.py
from operator import itemgetter

def set_query(val, d):
    visited = set()
    answer = 0
    while len(visited) < len(val):
        l = []
        for key in d.keys():
            if not key in visited:
                l.append([key,len(d[key]) * val[key - 1]])
        k = sorted(l, key=itemgetter(1))
        selected_key = l[0][0]
        visited.add(selected_key)
        answer += val[selected_key - 1]
        for i in range(len(val)):
            if i != selected_key - 1:
                if i + 1 in d[selected_key]:
                    val[i] += 1
                else:
                    val[i] -= 1
    return val, answer

n,q = map(int,input().split())

val = list(map(int,input().split()))
d = {}
for i in range(q):
    temp1, temp2 = map(int,input().split())
    if not temp1 in d:
        d[temp1] = set()
    if not temp2 in d:
        d[temp2] = set()
    d[temp1].add(temp2)
    d[temp2].add(temp1)
q = int(input())
while q:
    q-=1
    q_string = input()
    if q_string[0] == "+":
        q_string = q_string[2:]
        temp1, temp2 = map(int,q_string.split())
        if not temp1 in d:
            d[temp1] = set()
        if not temp2 in d:
            d[temp2] = set()
        d[temp1].add(temp2)
        d[temp2].add(temp1)
    elif q_string[0] == "-":
        q_string = q_string[2:]
        temp1, temp2 = map(int,q_string.split())
        if not temp1 in d:
            d[temp1] = set()
        if not temp2 in d:
            d[temp2] = set()
        d[temp1].remove(temp2)
        d[temp2].remove(temp1)
    else:
        val, answer = set_query(val, d)
        print(answer)
